### ISSUE TITLE
Using union types together with literal singleton types

### ISSUE THREAD

**Issue by joost-de-vries** [lampepfl/dotty#1495]
This is a follow up on a gitter conversation with @smarter 

I'm getting accustomed to some very handy uses of union types and literal singleton types in the Typescript part of our application. I've been trying this and some other TS uses of union types my beloved Scala with Dotty.

Things don't seem to work quite yet.

The most useful one is using _union types with literal singleton types_.  
At the moment that combination doesn't seem to work very much.

That's a pity because there's at least one use case that's very useful. It's possible to encode enumerations to constrain function arguments and assignments. 

``` scala
type Digit = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9   // seems to be interpreted as type Int

val digits: List[Digit] = List(-1, 0, 11)   // this compiles right now. 
```

It's something that has [just appeared in Typescript 2.0](https://blogs.msdn.microsoft.com/typescript/2016/08/30/announcing-typescript-2-0-rc/#user-content-more-literal-types) and it's really really useful.

I understood from @smarter that there's a concern around the performance of inferring return types:

> The main issue is that you can't always preserve singleton types and unions, otherwise any complex if/else if/../else will end up having a type like 10 | 32 | 50 | ..., doing subtyping checks on huge union types like this is likely to slow down the compiler
> my idea is that we should preserve unions of singleton types and not widen them only if the expected type is a union

Another feature I use a lot in Typescript is working with _non nullable types_. 
Instead of using an `Option[A]` you use a `A | Null` kind construction. Using a compiler flag or language feature import it would be possible to disallow null asignment.

``` scala
val foo: String = null // wouldn't compile

val bar: String | Null = null  // would compile
```

Currently the `Null` type seems to be thrown away.
[Here's the relevant TS doc](https://blogs.msdn.microsoft.com/typescript/2016/07/11/announcing-typescript-2-0-beta/#non-nullable-types)  
I think it's cool to have an alternative for the Option monad. But if it's hard to implement I'm perfectly fine with using `Option[A]`.

_Invoking a function on an object with a union type_ doesn't work right now with Dotty

``` scala
class Bird() {
    def fly()={}
    def layEggs()={}
}

class Fish() {
    def swim()={}
    def layEggs()={}
}

def getSmallPet(): Fish | Bird = if(new Random().nextBoolean) new Fish() else new Bird()

val pet = getSmallPet()   // returns Object right now
pet.layEggs() // This should compile shouldn't it?
pet.swim();    // and this shouldn't 
```

[The example is from Typescript docs](https://www.typescriptlang.org/docs/handbook/advanced-types.html)  
I guess this would be mostly useful for Scalajs since its compile target is very duck typing oriented.

For me and the Scala developers I've spoken to its the combination of union types and literal singleton types that's most valuable.

Btw I was amazed how quickly I got a response on gitter. On a Sunday! 😃

**Comment by smarter** [comment]
1. Union types with singleton types is something I'll try to look at in the next few weeks.
2. Currently `Null` is a subtype of every non-value class, non-nullable types are on our roadmap but require a substantial amount of work, especially if you want smooth interoperability with Java/Scala2
3. Your example works if you replace `val pet = getSmallPet()` with `val pet: Fish | Bird = getSmallPet()` or alternatively if you put `import scala.language.keepUnions` at the top of your file (note: I'm pretty sure this should be `import dotty.language.keepUnions`,  we seem to have a bug here). Currently we never infer union types without the `keepUnions` import but your example illustrates a case where we might want to relax this rule, this might be hard to do since you need to precisely keep track of the source of your inferred type.

**Comment by felixmulder** [comment]
To add to this discussion, it doesn't seem like this is possible anymore:

```scala
object Monday
object Tuesday
object Saturday
object Sunday

type Weekday = Monday.type | Tuesday.type
type Weekend = Saturday.type | Sunday.type
type AnyDay = Weekday | Weekend

(Monday: AnyDay) match {
  case _: Weekend => ... // matches this case!
  case _ => ...
}
```

Ping @smarter

**Comment by smarter** [comment]
@felixmulder Your example works fine on the latest master, probably fixed by #1491

**Comment by felixmulder** [comment]
The example in question was:

```scala
object Test {
  object Monday
  object Tuesday
  object Wednesday
  object Thursday
  object Friday

  object Saturday
  object Sunday

  type Weekday = Monday.type | Tuesday.type | Wednesday.type | Thursday.type | Friday.type
  type Weekend = Saturday.type | Sunday.type
  type AnyDay  = Weekday | Weekend

  def main(args: Array[String]): Unit = {
    println("Monday is weekday: " + Monday.isInstanceOf[Weekday])
    println("Saturday is weekend: " + Saturday.isInstanceOf[Weekend])
    println("Sunday is weekday: " + Sunday.isInstanceOf[Weekday])

    (Monday: AnyDay) match {
      case _: Weekend => println("shouldn't match")
    }
  }
}
```

Which before #1491 would yield:

```
Monday is weekday: true
Saturday is weekend: true
Sunday is weekday: true
shouldn't match
```

**Comment by joost-de-vries** [comment]
See #1551

**Comment by joost-de-vries** [comment]
[This comment](https://contributors.scala-lang.org/t/union-type-widening/1069/3) by Felix Smarter describes the performance concerns in more detail.

**Comment by joost-de-vries** [comment]
Closing this since it is unlikely to be changed, witness #1532

**Comment by joost-de-vries** [comment]
Fixed by #6299 🎉
